src/topology/dense_embedding.lean

```diff
@@ -0,0 +1,245 @@
+/-
+Copyright (c) 2019 Reid Barton. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johannes Hölzl, Mario Carneiro, Patrick Massot
+-/
+import topology.maps topology.separation
+
+/-!
+# Dense embeddings
+
+This file defines three properties of functions:
+
+`dense_range f`      means `f` has dense image;
+`dense_inducing i`   means `i` is also `inducing`;
+`dense_embedding e`  means `e` is also an `embedding`.
+
+The main theorem `continuous_extend` gives a criterion for a function
+`f : X → Z` to a regular (T₃) space Z to extend along a dense embedding
+`i : X → Y` to a continuous function `g : Y → Z`. Actually `i` only
+has to be `dense_inducing` (not necessarily injective).
+
+-/
+
+noncomputable theory
+
+open set filter lattice
+open_locale classical
+
+variables {α : Type*} {β : Type*} {γ : Type*} {δ : Type*}
+
+section dense_range
+variables [topological_space β] [topological_space γ] (f : α → β) (g : β → γ)
+
+/-- `f : α → β` has dense range if its range (image) is a dense subset of β. -/
+def dense_range := ∀ x, x ∈ closure (range f)
+
+lemma dense_range_iff_closure_eq : dense_range f ↔ closure (range f) = univ :=
+eq_univ_iff_forall.symm
+
+variables {f}
+
+lemma dense_range.comp (hg : dense_range g) (hf : dense_range f) (cg : continuous g) :
+  dense_range (g ∘ f) :=
+begin
+  have : g '' (closure $ range f) ⊆ closure (g '' range f),
+    from image_closure_subset_closure_image cg,
+  have : closure (g '' closure (range f)) ⊆ closure (g '' range f),
+    by simpa [closure_closure] using (closure_mono this),
+  intro c,
+  rw range_comp,
+  apply this,
+  rw [(dense_range_iff_closure_eq f).1 hf, image_univ],
+  exact hg c
+end
+
+/-- If `f : α → β` has dense range and `β` contains some element, then `α` must too. -/
+def dense_range.inhabited (df : dense_range f) (b : β) : inhabited α :=
+⟨begin
+  have := exists_mem_of_ne_empty (mem_closure_iff.1 (df b) _ is_open_univ trivial),
+  simp only [mem_range, univ_inter] at this,
+  exact classical.some (classical.some_spec this),
+ end⟩
+
+lemma dense_range.nonempty (hf : dense_range f) : nonempty α ↔ nonempty β :=
+⟨nonempty.map f, λ ⟨b⟩, @nonempty_of_inhabited _ (hf.inhabited b)⟩
+end dense_range
+
+/-- `i : α → β` is "dense inducing" if it has dense range and the topology on `α`
+  is the one induced by `i` from the topology on `β`. -/
+structure dense_inducing [topological_space α] [topological_space β] (i : α → β)
```

Comment:
Should we use `old_structure_cmd` here?

Reply:
I think I tried, and decided to stick to new, but I don't remember why. Maybe this is something @rwbarton could experiment during refactoring.

Reply:
I see on Zulip that Reid prefers to keep such experiment for other PRs, so let's mark this as resolved.